tests/check_framework/test_templates.py

```diff
@@ -1,59 +1,74 @@
+from django.template.backends.base import BaseEngine
+from django.template.backends.django import E002, W003
 from django.test import SimpleTestCase
 from django.test.utils import override_settings
 
 
-class CheckTemplateSettingsAppDirsTest(SimpleTestCase):
-    TEMPLATES_APP_DIRS_AND_LOADERS = [
-        {
-            "BACKEND": "django.template.backends.django.DjangoTemplates",
-            "APP_DIRS": True,
-            "OPTIONS": {
-                "loaders": ["django.template.loaders.filesystem.Loader"],
-            },
-        },
-    ]
+class TemplateEngineCheckTests(SimpleTestCase):
+    def test_check_templates_aggregates_checks_from_all_registered_engines(self):
+        error1 = Error("1")
+        error2 = Error("2")
+        error3 = Error("3")
 
-    @override_settings(TEMPLATES=TEMPLATES_APP_DIRS_AND_LOADERS)
-    def test_app_dirs_and_loaders(self):
-        """
-        Error if template loaders are specified and APP_DIRS is True.
-        """
-        self.assertEqual(check_setting_app_dirs_loaders(None), [E001])
+        def get_engines():
+            class Engine1(BaseEngine):
+                def check(self, **kwargs):
+                    return [error1]
 
-    def test_app_dirs_removed(self):
-        TEMPLATES = deepcopy(self.TEMPLATES_APP_DIRS_AND_LOADERS)
-        del TEMPLATES[0]["APP_DIRS"]
-        with self.settings(TEMPLATES=TEMPLATES):
-            self.assertEqual(check_setting_app_dirs_loaders(None), [])
+            class Engine2(BaseEngine):
+                def check(self, **kwargs):
+                    return []
 
-    def test_loaders_removed(self):
-        TEMPLATES = deepcopy(self.TEMPLATES_APP_DIRS_AND_LOADERS)
-        del TEMPLATES[0]["OPTIONS"]["loaders"]
-        with self.settings(TEMPLATES=TEMPLATES):
-            self.assertEqual(check_setting_app_dirs_loaders(None), [])
+            class Engine3(BaseEngine):
+                def check(self, **kwargs):
+                    return [error2, error3]
+
+            class Engine4(BaseEngine):
+                pass
+
+            params = {"DIRS": [], "APP_DIRS": False, "NAME": "engine"}
+            return [Engine1(params), Engine2(params), Engine3(params), Engine4(params)]
+
+        with mock.patch("django.template.engines.all", get_engines):
+            errors = check_templates(None)
+            self.assertListEqual(errors, [error1, error2, error3])
+
+    def test_djangotemplates_check_aggregates_backend_checks(self):
+        error1 = Error("1")
+        error2 = Error("2")
+        django_engine = engines["django"]
+        with (
+            mock.patch.object(
+                django_engine, "_check_string_if_invalid_is_string", lambda: [error1]
+            ),
+            mock.patch.object(
+                django_engine,
+                "_check_for_template_tags_with_the_same_name",
+                lambda: [error2],
+            ),
+        ):
+            self.assertListEqual(django_engine.check(), [error1, error2])
```

Comment:
I really appreciate that you’ve tested the top-level function, but I think the tests are too complicated, especially given how simple the function is. Also, mocking should be avoided -  `override_settings` can instead be used to swap the engine classes, if they are defined outside of the test method.

Below is an untested example test that I think would suffice for everything:

```
class ErrorEngine(BaseEngine):
    def check(self, **kwargs):
        return [Error("Example")]


class CheckTemplatesTests(SimpleTestCase):
    @override_settings(
        TEMPLATES=[
            {"BACKEND": f"{__name__}.{ErrorEngine.__qualname__}"},
            {"BACKEND": f"{__name__}.{ErrorEngine.__qualname__}"},
        ]
    )
    def test_errors_aggregated(self):
        errors = check_templates(None)
        self.assertEqual(
            errors,
            [Error("Example")] * 2,
        ]
```

Note you don’t normally need `assertListEqual` or other typed functions, unittest maps `assertEqual` to type-specific functions: https://docs.python.org/3.12/library/unittest.html#type-specific-methods